packages/engine/src/datastore/schema/field_spec/mod.rs

```diff
@@ -79,142 +33,83 @@ pub enum FieldScope {
     Agent,
 }
 
-/// Defines the source from which a Field was specified, useful for resolving clashes
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub enum FieldSource {
-    Engine,
-    Package(PackageName),
-}
-
-impl FieldSource {
-    /// A unique static identifier of the package source, used in building Keys for fields
-    pub fn unique_id(&self) -> Result<String> {
-        match self {
-            FieldSource::Engine => Ok("0".into()),
-            FieldSource::Package(package_name) => Ok(package_name
-                .get_id()
-                .map_err(|err| Error::from(err.to_string()))?
-                .to_string()),
-        }
-    }
-}
-
-#[derive(Debug, Clone, Eq, PartialEq, Hash)]
-pub struct FieldKey(String);
-
-impl FieldKey {
-    /// Returns the key as string
-    pub fn value(&self) -> &str {
-        &self.0
-    }
-
-    /// Returns a string as key
-    pub(in crate::datastore) fn new(key: &str) -> Self {
-        Self(key.to_string())
-    }
-
-    /// Create a new agent scoped `FieldKey`
-    ///
-    /// Builds a `FieldKey` from a given name.
+impl FieldScope {
+    /// Create a new `FieldKey` from this scope with the provided `name` and the specified source.
     ///
     /// # Errors
     ///
     /// - Returns [`Error`] if name starts with [`PRIVATE_PREFIX`] or [`HIDDEN_PREFIX`]
-    #[inline]
-    pub fn new_agent_scoped(name: &str) -> Result<Self> {
```

Comment:
We've completely lost this signature with this new approach and we're no longer clearly giving the different requirements for different scopes of FieldKeys. I'm not sure how I feel about us treating FieldScope as a factory for FieldKeys, it feels unintuitive to not be able to create one from the FieldKey struct.

Reply:
I made `FieldSource` a trait, so those methods could be moved to the new crate as well. Specifically this changes got revertwed.